src/hpke.rs

```diff
@@ -0,0 +1,610 @@
+// Copyright (C) Nitrokey GmbH
+// SPDX-License-Identifier: Apache-2.0 or MIT
+
+use crate::StagingBackend;
+
+use trussed::{
+    config::MAX_SERIALIZED_KEY_LENGTH,
+    key,
+    serde_extensions::ExtensionImpl,
+    service::Filestore,
+    store::keystore::Keystore,
+    types::{KeyId, Message},
+    Bytes,
+};
+use trussed_hpke::*;
+
+type HkdfSha256 = hkdf::Hkdf<sha2::Sha256>;
+type HkdfSha256Extract = hkdf::HkdfExtract<sha2::Sha256>;
+
+use rand_core::{CryptoRng, RngCore};
+use salty::agreement as x25519;
+
+const X25519_KEM_SUITE_ID: &[u8] = b"KEM\x00\x20";
+const X25519_HKDF_SHA256_CHACHA20_POLY1305_HPKE_SUITE_ID: &[u8] = b"HPKE\x00\x20\x00\x01\x00\x03";
+
+fn labeled_extract(
+    suite_id: &[u8],
+    salt: &[u8],
+    label: &[u8],
+    ikm: &[u8],
+) -> (HkdfSha256, [u8; 32]) {
+    let mut extract_ctx = HkdfSha256Extract::new(Some(salt));
+    extract_ctx.input_ikm(b"HPKE-v1");
+    extract_ctx.input_ikm(suite_id);
+    extract_ctx.input_ikm(label);
+    extract_ctx.input_ikm(ikm);
+    let (prk, hkdf) = extract_ctx.finalize();
+    (hkdf, prk.into())
+}
+
+fn labeled_expand(
+    suite_id: &[u8],
+    prk: &HkdfSha256,
+    label: &[u8],
+    info: &[u8],
+    buffer: &mut [u8],
+) -> Result<(), hkdf::InvalidLength> {
+    let Ok(l): Result<u16, _> = buffer.len().try_into() else {
+        return Err(hkdf::InvalidLength);
+    };
+    prk.expand_multi_info(
+        &[&l.to_be_bytes(), b"HPKE-v1", suite_id, label, info],
+        buffer,
+    )
+}
+
+fn extract_and_expand(dh: x25519::SharedSecret, kem_context: &[u8]) -> [u8; 32] {
+    let (prk, _) = labeled_extract(X25519_KEM_SUITE_ID, b"", b"eae_prk", &dh.to_bytes());
+    let mut shr = [0; 32];
+    labeled_expand(
+        X25519_KEM_SUITE_ID,
+        &prk,
+        b"shared_secret",
+        kem_context,
+        &mut shr,
+    )
+    .expect("Length of shr is known to be OK");
```

Comment:
replace `expect` with `error!` + `unwrap`?

Reply:
It’s not necessary, but if we would ever end up using the implementation e. g. for the Passkey, we will probably not remember to fix this.  On the other hand, maybe the `expect` optimization will have improved by then. ;)